include/st_string.h

```diff
@@ -1200,6 +1256,22 @@ namespace ST
         }
 #endif // defined(ST_ENABLE_STL_FILESYSTEM) && defined(ST_HAVE_CXX17_FILESYSTEM)
 
+#ifdef ST_ENABLE_COREFOUNDATION
+        ST_NODISCARD ST_RETURNS_RETAINED
+        CFStringRef to_CFString() const
+        {
+            return CFStringCreateWithCString(kCFAllocatorDefault, c_str(), kCFStringEncodingUTF8);
+        }
+
+#ifdef __OBJC__
+        ST_NODISCARD
+        NSString* to_NSString() const
+        {
+            return ST_BRIDGED_TRANSFER_CAST(NSString*)to_CFString();
```

Comment:
My thinking was this:

`to_CFString` calls a `CFStringCreate*` function, which means the result needs to be freed with `CFRelease` by the caller at a later point. Hence that string is retained (it is returned with a reference count of 1).

However, in Objective-C(++), I believe assignments will increase the reference count. Since the underlying CFString starts out with a reference count of 1, storing the result of `to_NSString()` in an `NSString*` variable would end up incrementing it to 2.

```c++
CFStringRef str = stStr.to_CFString(); // refCnt 1
// Do stuff with str
CFRelease(str); // refCnt 0 -> object is freed
```

```objc
// Without ARC:
NSString* str = stStr.to_NSString(); // refCnt 1
// Do stuff with str

// Approximating ARC:
@autoreleasepool {
  NSString* str2 = str; // refCnt 2
  // Do stuff with str2
} // refCnt 1

[str release]; // refCnt 0 -> object is freed
```

Reply:
No, assigning an object to a plain variable won't change its refcount. (Assigning to a *property* usually retains the object - but it also may not, or it might make a copy, or do something else entirely, because properties are actually getter/setter methods in disguise.)

In your example: if `to_NSString()` returns an unretained object, then you don't need the `[str release]` at the end, because `str` was never retained for you. If `to_NSString()` returns a retained object, then you do need the `[str release]`. In either case, the `@autoreleasepool` and extra variable in the middle don't affect anything.

Now that I read it again, I think the current implementation actually has inconsistent behavior depending on whether ARC is enabled... With ARC disabled, your `ST_BRIDGED_TRANSFER_CAST` macro does a plain cast and nothing else, so the method returns a retained object. With ARC enabled, `ST_BRIDGED_TRANSFER_CAST` causes an implicit (auto?)release, so the method returns an unretained object.

With `ST_RETURNS_RETAINED` and a simple bridging cast, it would be consistent. Both with and without ARC, a retained object would be returned. Without ARC, the object has to be manually released in the calling code. With ARC, the calling code releases it implicitly, because of the `ST_RETURNS_RETAINED` annotation.